### PR TITLE
Fix deploy scripts for handling composer.json version and version string validation

### LINKED ISSUE #3026: Improve deploy scripts to correctly handle patch versions and composer.json versioning.
Currently, our deploy script treats `x.x.0` versions as patch versions, giving this warning
```
The version you entered (3.2.0) looks like a patch version. Since this version will be deployed to WordPress.org, it will become the latest available version. Are you sure you want that (no will abort)?: [y/N]
```
versions ending with .0 are not patch versions.

The script also accepts `x.x` as a valid version number, so  `3.2` is a valid version, which it isn't, npm build will fail if the package.json version is `x.x`
image

So we need to reject this pattern and only accept `x.x.x`

-------
Edit by Darren: I've repurposed this issue to address various problems needing fixed in the deploy scripts including the above that were identified as a part of the 3.2.0 deploy process:

- `npm run deploy` incorrectly considers `x.x.0` a patch version.
- any script involving versioning updates should enforce a `x.x.x` format (the original problem prompting this issue).
- `npm run package-plugin:*` scripts don't build a plugin with the correct autoload map (need to do a temporary version bump in the composer.json, then execute `composer dump-autoload` before zipping the file).

### PR DESCRIPTION
**Pull request by nerrad**
Fixes #3026 

During the release of 3.2.0 there were a number of issues that popped up for @senadir around the [recent changes for handling correct autoload class map generation](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2949) using `npm run package-plugin/*` variations etc. There were also some issues related to the `npm run deploy` script and validation happening around the version string provided by user prompt.

In this pull:

- `npm run deploy` will now automatically coerce incorrect provided version strings to a semantic value that we use for releases. Concretely, this means:
    - If just the major version is provided, it will be automatically fixed. Example: `4` would become `4.0.0`.
    - If just the minor version is provided, it will be automatically fixed. Example: `3.3` would become `3.3`.
    - Minor patch versions are respected. Example: `3.3.1` would stay the same.
- `npm run deploy` will only warn on patch versions for _actual_ patch version. In other words, `3.3.0` would not be warned on because that's actually a minor release. Whereas `3.3.1` would receive the wp.org release warning.
- `npm run deploy` will no longer _commit_ the `composer.json` version bump but still commits the autoload class map generated with `version` in `composer.json`. For releases, `composer.json` doesn't actually need the version we just need to make sure we're generating the classmap with it.
- `npm run package-plugin/*` variations will now correctly account for generating the correct classmap in the built zip. This is done by updating the version in `composer.json` before the classmap is generated. After the zip is built, the version property is removed from `composer.json` and then the classmap is regenerated for the development environment.

## Testing

Testing these changes requires commenting out some lines to avoid unintentionally doing deploys! 

### For `npm run deploy` test

- in the file `./bin/github-deploy.sh`, comment out any line that is a `git` command _except_ `git checkout -- composer.json`.
- comment out the `hub` related commands.
- comment out the final `composer dump-autoload` in the file (near the end of the file)
* [ ]  execute `sh ./bin/github-deploy.sh` and answer with different variations of version strings and verify the correct versions are updated in files for the string given.
- [ ] verify the correct classmap generation has occurred. You do this by opening up `vendor/composer/jetpack_autoload_classmap.php`, search for anything "Blocks" related, and check that the `version` property for the class uses the string you passed in as the version for the script run. **Note:** it's expected that the version string in jetpack has a trailing `.0`in version strings.
- re-enable the final `composer dump-autoload` in the file you commented out earlier.
- [ ] re-execute the script and verify the correct classmap has occurred. This time the classmap should have the branch name for the `version` string.
- [ ] verify that `composer.json` does not have a "version" property after running the script.

### For `npm run package-plugin` test

Note: before you run each of these scripts you either have to [comment out this code](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/0fdf262bfb350d91a4fd55faf7900397f21e52c7/bin/build-plugin-zip.sh#L65-L81) or commit your changes locally (and don't push) to avoid the dirty repo bail in `build-plugin-zip.sh`.

- in the file `./bin/build-plugin-zip.sh`, comment out the final `composer dump-autoload` (near the end of the file).
* [ ] execute `npm run package-plugin:deploy` (you can do whatever package-plugin command you want, but this is the critical one) and verify that the generated autoloader class map (same location as previous test) has the expected "version" property for Blocks classes. The expected value should match whatever version is in `package.json`.
- [ ] verify that `composer.json` does not have a "version" property after running the script.
- re-enable the final `composer dump-autoload` in the file you commented out earlier.
* [ ] re-execute the `npm run package-plugin:deploy` command. This time the generated autoloader classmap in your local directory should be using the branch name as the "version" property value for blocks classes. However, in the **built zip**, the autoloader classmap should have the same value for the "version" property as the version property in `package.json`.

### CODE REVIEW

**Review by haszari — COMMENTED** [review]
Thanks @nerrad - great to see our release tooling getting more robust ❤️ 

I've done my best to test this, but I hit a snag with the `package-plugin` tests - my tree is dirty (because of hacked scripts), so it bails (see logs below). I might have misunderstood the test instructions – it's quite complex, and because I don't have a solid understanding of our scripts, I was blindly following instructions. So you might want a review from someone more familiar, or hit me back with more info about how to verify this safely. 

I did manage to test `npm run deploy` and it seemed to be working correctly (based on your instructions). I found one small issue - trailing `.` in version number is preserved. I guess it would be nice to enforce `x.y.z` format exactly or bail with an error, but this is probably not urgent – we can document the requirement for `x.y.z` format.\

I have a bunch of thoughts/feedback though – up to you whether we need to make changes in this PR or just ideas for enhancements in future.

When reviewing code, I skimmed over most of this, especially the regex logic. If you want a good code-review of the changes to the scripts maybe ping me back or get another review. 

### Consider updating release process doc
Does this change to the scripts change any of the release process - should the docs get updated in this PR? I think maybe there's a new expectation that `composer.json` is updated with (dev?) version after release. (Based on discussions in slack)

Could also potentially clarify the inputs/outputs/side effects of scripts in docs to help us all better understand how they work – they are mostly a black box for me right now. Release docs might be one way to help this, or could have a separate doc for the npm scripts related to build => release. A lot of complexity has evolved here :)

### Clarify "incorrect autoload map" issue & impact
> `npm run package-plugin/*` variations will now correctly account for generating the correct classmap in the built zip.

I don't understand the actual failure mode / problem for this aspect of the PR. Might be a good idea to clarify that in the issue or PR for posterity and to spread knowledge. I.e. package-plugin was previously building a plugin with an "incorrect" autoload map - what does this mean, what is the effect, how is the plugin broken? (Note - I just want to understand this better, I'm not questioning the need or the fix!)

### Testability of scripts
Commenting out lines of code in deploy scripts to test is quite unnerving, especially if you're doing it "blind" (i.e. you don't have a solid understanding of how the scripts work).

I can think of two possible improvements here – maybe worth iterating on this over time:

1. Add a `--dry-run` param to potentially dangerous scripts. This way, when we need to test a script that deploys, or commits, or pushes (etc), we can do so without having to hack code. 
2. Refactor scripts so high-impact and "safe" functionality can be run separately. (This could help power a `--dry-run` option.)

Note there's no silver bullet here – the only way to really ensure these scripts are reliable and robust is to use them for releases regularly. I'm advocating for refactoring and `--dry-run` to help us all understand the scripts more, and to make testing/review easier as a secondary benefit.

### WooRelease
I realise we're committed to our tooling for the time being. However I think it's worth mentioning - the more we invest in our custom tooling, the more overhead for us as a team (as we're the only clients).

Are our needs special, i.e. we need custom tooling? Or, is it that our tooling has more features not (yet) in WooRelease? Ideally when we touch our tooling we iterate closer to sharing more of it – i.e. log enhancement requests for WooRelease, or code our custom stuff in a way that it could be adapted to WooRelease in future. 

Totally not a blocker!!

## Testing…

### `npm run deploy` test
Note: I tested this in reverse order by mistake. My first test had `composer dump-autoload` NOT commented. Once I realised this, I confirmed version numbers in  were correct in `jetpack_autoload_classmap` - dev version number (`composer dump-autoload` uncommented), then real version number (`composer dump-autoload` commented).

The only issue I saw was trailing `.` in version number wasn't normalised correctly (see log below for an example).

Note in `jetpack_autoload_classmap.php` the version number has 4 fields - i.e. trailing 0 - assuming that's expected.

### `npm run package-plugin` test

* [ ] re-execute the `npm run package-plugin:deploy` command. 

When I ran this it bailed because of dirty source files (see below) - is this expected? Maybe I need to commit the script hacks to a temporary branch?

##### logs (aka screenshots)
```
// 3.6.78. not converted to X.Y.Z format
Please enter the version number to tag, for example, 1.0.0:
3.6.78.

// …

The version is set as  3.6.78.  and the next step will be to bump all the version strings in relevant files.
```

```
// version in classmap - 4 fields - looking good
  'Automattic\\WooCommerce\\Blocks\\Payments\\PaymentResult' => array(
    'version' => '3.6.2.0',
    'path'    => $baseDir . '/src/Payments/PaymentResult.php'
  ),
```

```
// can't run package-plugin:deploy with dirty clone
> ./bin/build-plugin-zip.sh -z


💃 Time to build a WooCommerce Blocks ZIP 🕺

HEAD detached at origin/fix/3026-fix-deploy-scripts
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  modified:   bin/build-plugin-zip.sh
  modified:   bin/github-deploy.sh
  modified:   package.json
  modified:   readme.txt
  modified:   src/Package.php
  modified:   woocommerce-gutenberg-products-block.php

Untracked files:
  (use "git add <file>..." to include in what will be committed)
  tests/php/Domain/Services/TestCreateAccount.php

no changes added to commit (use "git add" and/or "git commit -a")

ERROR: Cannot build plugin zip with dirty working tree. ☝️
    Commit your changes and try again.
```

File: bin/build-plugin-zip.sh
Comment (nerrad) [reply bin/build-plugin-zip.sh:128]: left over from debugging, good catch 👍

File: bin/github-deploy.sh
Comment (nerrad) [reply bin/github-deploy.sh:110]: The current configured workflow runs a WordPress deploy whenever a GitHub _release_ is created (differs from just a `tag`).  Earlier in the flow, the user is asked if this release is being deployed to WordPress. The problem that can happen is if the user indicates they are releasing a patch version to WordPress and we already have a higher version available on WordPress, this will result in the available version to be bumped down to the patch version.

So to be clear, this question only gets asked when:

- user has answered yes to deploying this on WordPress.org (Yes means a GitHub release gets created, no means just a GitHub tag is created).
- The version the user provided is detected as a patch version. It meets the criteria of `x.x.[>0]`

_So why ask the question, why not just abort?_

The question is asked because there _may_ be cases where the patch version _would_ be the latest version on WP.org and we _do_ want to release there.

What I can do is improve the first question about WordPress version and provide more context there (so the user knows when to answer yes or no).

File: bin/github-deploy.sh
Comment (nerrad) [reply bin/github-deploy.sh:185]: All good points - I'll create an todo (in internal "document all the things post") to improve documentation around this. A lot of the complexity here is due to usage of jetpack autoloader and the feature plugin being available in WooCommerce core as a package but _also_ as a regular WordPress plugin. So there is need to ensure that when the feature plugin version is greater than the version bundled with WooCommerce core the feature plugin classes are used instead of the package classes in Woo Core.

A lot of the fixes made in these scripts is to remove the mental overhead of having to remember what needs done for version management but I agree for _maintainance_ of these scripts, it'd be good to document what's happening.

**Comment by nerrad** [comment]
Thanks for your solid review Rua. I've addressed your feedback in the inline review and I'll address your other comments below. Please keep in mind that I'm trying to get the immediate problems with the deploy script fixed before the next WC Blocks release so it can be used reliably and we can avoid the issues Nadir encountered in the last release. So keeping that in mind, let's consider what can be done separately from this pull vs what definitely needs addressed in this pull before merge. I do appreciate that you've already prefaced your additional feedback with that consideration.

> I've done my best to test this, but I hit a snag with the package-plugin tests - my tree is dirty (because of hacked scripts), so it bails (see logs below). I might have misunderstood the test instructions – it's quite complex, and because I don't have a solid understanding of our scripts, I was blindly following instructions. So you might want a review from someone more familiar, or hit me back with more info about how to verify this safely.

Sorry about that, ya I forgot about the dirty working tree bail in `build-plugin-zip.sh`. I've updated the testing instructions with how to work around that for the purpose of the tests.

> I did manage to test npm run deploy and it seemed to be working correctly (based on your instructions). I found one small issue - trailing . in version number is preserved. I guess it would be nice to enforce x.y.z format exactly or bail with an error, but this is probably not urgent – we can document the requirement for x.y.z format.\

Can you clarify more about this please? Do you mean if someone explicitly indicates `3.1.3.` in the requested version that is preserved? If yes, that's a pre-existing issue (outside of this branch there's even more accepted formats) so I'd prefer it's handled separately. 

> Does this change to the scripts change any of the release process - should the docs get updated in this PR? I think maybe there's a new expectation that composer.json is updated with (dev?) version after release. (Based on discussions in slack)

Good point, I think I intended to do some changes in the release docs in this pull but forgot. I've added a few changes. In a pending automation I'll be adding soon for creating the release pull request, there will be a number of other updates/changes needed as well because of what the automation will take care of so I've limited what changes I make now to just what's necessary.

> I don't understand the actual failure mode / problem for this aspect of the PR. Might be a good idea to clarify that in the issue or PR for posterity and to spread knowledge. I.e. package-plugin was previously building a plugin with an "incorrect" autoload map - what does this mean, what is the effect, how is the plugin broken? (Note - I just want to understand this better, I'm not questioning the need or the fix!)

I updated the pull description to add a link to [this pull](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2949) which adds more context.

> Commenting out lines of code in deploy scripts to test is quite unnerving, especially if you're doing it "blind" (i.e. you don't have a solid understanding of how the scripts work).
> 
> I can think of two possible improvements here – maybe worth iterating on this over time:

I agree, and actually thought of implementing a dry run when I was working on this, but decided to pragmatically focus on fixing the primary issues first to facilitate the usability of the script for the next Blocks release. I'm definitely open to implementing your suggestions in future iterations (if you can open an issue that'd be great).

> Are our needs special, i.e. we need custom tooling? Or, is it that our tooling has more features not (yet) in WooRelease? Ideally when we touch our tooling we iterate closer to sharing more of it – i.e. log enhancement requests for WooRelease, or code our custom stuff in a way that it could be adapted to WooRelease in future.

WooRelease is a great tool. However, from what I've seen, it currently has a lot more than what we have need of in what we are doing and _currently_ the effort to update our existing scripts and GitHub actions outweigh the effort needed to try using WooRelease and make any changes needed (either via custom commands or upstream pulls). Another consideration is WooRelease is a `php` based tool and thus needs a php environment to run in. Our currently tooling uses shell/node and is inherently more friendly to _quicker_  integration with GitHub actions (which we're using more and more of for our automation needs).

I'm not opposed to investigating and implementing usage of WooRelease in the future but given the resources currently available to our team I'm pragmatically making the decision to the minor adjustments needed to our _existing_ tooling which is already sufficient for our needs in this repo (and arguably doing more than a straight up switch to WooRelease would accomplish for us at the moment).

Where WooRelease is a _really_ great option is for projects that don't have any automation (or very basic) automation tools/scripts at their disposal. 

> Note in jetpack_autoload_classmap.php the version number has 4 fields - i.e. trailing 0 - assuming that's expected.

Yup that's expected - I've updated the pull description testing notes.

File: docs/releases/handling-releases.md
Comment (haszari) [inline docs/releases/handling-releases.md:161]: Might be even clearer to remove `update` since this is automated. Something like: 

```
-   For _major_ & _minor_ releases, confirm version on `main` is correct and has dev suffix, e.g. [`2.6-dev`](https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/e27f053e7be0bf7c1d376f5bdb9d9999190ce158). 
**Note:** this is automated by the release deploy script, just verify that the versions have been updated correctly.
```

File: bin/github-deploy.sh
Comment (haszari) [inline bin/github-deploy.sh:80]: Minor typo, missing `than`: `If the version on WordPress.org is greater than ${VERSION}`

**Review by haszari — APPROVED** [review]
New test instructions worked well - I committed my testing hacks to a temporary branch. All the version numbers and checks worked so this is looking 👍

> Do you mean if someone explicitly indicates 3.1.3. in the requested version that is preserved? If yes, that's a pre-existing issue (outside of this branch there's even more accepted formats) so I'd prefer it's handled separately.

Yes, correct - user enters version like `1.3.5.` with trailing dot when prompted. I agree, this does not need to be fixed in this PR :)

> I agree, and actually thought of implementing a dry run when I was working on this, but decided to pragmatically focus on fixing the primary issues first to facilitate the usability of the script for the next Blocks release. I'm definitely open to implementing your suggestions in future iterations (if you can open an issue that'd be great).

Sounds good – logged here: #3059 

Feel free to take over that issue, especially if you have specific ideas for pragmatic solutions we can implement soon. (The issue is currently more "vision" if you know what I mean.)

> I'm not opposed to investigating and implementing usage of WooRelease in the future but given the resources currently available to our team I'm pragmatically making the decision to implement the minor adjustments needed to our existing tooling which is already sufficient for our needs in this repo (and arguably doing more than a straight up switch to WooRelease would accomplish for us at the moment).

Makes sense to me - thanks for explaining more about this.

File: docs/releases/handling-releases.md
Comment (nerrad) [reply docs/releases/handling-releases.md:161]: As an aside, are you aware of the ["suggest" feature in GitHub](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request#adding-line-comments-to-a-pull-request)? You can directly suggest the change in pull request review and then the author of the pull request just has to accept it to merge.

**Pull request merged**
